### PR TITLE
Check HAVE_CONFIG_H before including "config.h"

### PR DESCRIPTION
**Pull request by pherl**
The first step of introducing "config.h"-less BUILD.

### CODE REVIEW

**Comment by pherl** [comment]
@xfxyjwf  for review

**Comment by xfxyjwf** [comment]
Where is this HAVE_CONFIG_H macro defined?

**Comment by pherl** [comment]
It's defined by automake/autoconf I believe. For builds other that the automake, we need to
manually write a config.h and pass -DHAVE_CONFIG_H.

Checked the places where config.h is introduced:
1) to generate pbconfig.h for hash maps/setes, which will be no longer needed.
2) in common.cc for pthread related stuffs. We have macro checks for windows. Thus we need to manually pass -DHAVE_PTHREAD for bazel.
3) gzip streams and tests, which relies on HAVE_ZLIB

**Comment by xfxyjwf** [comment]
For the config.h, we can:
  1. Always generate config.h. This is what we currently do.
  2. Do not generate config.h. Instead of generating config.h, the build tools can just pass macros to the compiler directly (e.g., through CXXFLAGS when building with autoconf/automake).
  3. Optionally generate config.h. The code can compile both with/without config.h.

I am confused by what you are trying to do in this PR. It seems you are trying to go with option 3). But if the code is required to compile without config.h anyway, why don't we just get rid of the config.h file? (i.e., 2 is strictly better than 3 in my opinion).

**Comment by pherl** [comment]
SGTM. The only downside for option 2 is that if in the future we add need more macros from autoconf, passing -DXXX may make the output harder to read.

**Comment by pherl** [comment]
Just to capture the discussion we had yesterday.
We will remove config.h includes from srcs, rather than checking HAVE_CONFIG_H:
- Change Makefile.am to pass in HAVR_ZLIB manually
- HAVE_PTHREAD always needs to be set on non-windows platforms, so remove the ifdef on it.

The config.h will still be generated, because
1. otherwise autoconf will pass those macros from -DXXXX in CXXFLAGS, which spams the logs.
2. for users porting the code to other platforms, it is easier to see what macros are potentially needed by looking at the config.h file.

**Comment by pherl** [comment]
Removed the includes and fixed cmake as well, PTAL.

File: src/Makefile.am [hunk 1]
```diff
@@ -1,20 +1,29 @@
 ## Process this file with automake to produce Makefile.in
 
+
```
Comment (xfxyjwf) [inline src/Makefile.am]: Remove this extra new-line?

**Comment by xfxyjwf** [comment]
LGTM

You might want to squash the commits into one.

**Pull request merged**